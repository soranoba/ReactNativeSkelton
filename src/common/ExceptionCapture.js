/**
 * @author Hinagiku Soranoba <soranoba@gmail.com>
 * @license MIT
 */
import React from 'react';
import { Platform } from 'react-native';
import PropTypes from 'prop-types';

/**
 * It is a Component like ErrorBounday, but it catch javascript exceptions unlike ErrorBounday.
 * You should not use more than one it.
 */
export default class ExceptionCapture extends React.Component {
  static propTypes = {
    children: PropTypes.node,
    component: PropTypes.func.isRequired,
  };
  constructor(props) {
    super(props);
    this.state = { error: null, hasError: false, originalGlobalHandler: null };
  }
  componentDidMount() {
    var originalGlobalHandler = null;
    const handler = error => {
      this.setState({ error, hasError: true });
    };

    if (Platform.OS === 'web') {
      originalGlobalHandler = window.onerror;
      window.onerror = (message, url, line, col, error) => {
        handler(error);
      };
    } else {
      originalGlobalHandler = ErrorUtils.getGlobalHandler();
      ErrorUtils.setGlobalHandler(handler);
    }

    // NOTE: To avoid `react/no-did-mount-set-state`
    (() => {
      this.setState({ originalGlobalHandler });
    })();
  }
  componentWillUnmount() {
    ErrorUtils.setGlobalHandler(this.state.originalGlobalHandler);
  }
  render() {
    if (this.state.hasError) {
      const Component = this.props.component;
      return <Component error={this.state.error} />;
    }
    return this.props.children;
  }
}
