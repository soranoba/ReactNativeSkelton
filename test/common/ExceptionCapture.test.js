import React from 'react';
import renderer from 'react-test-renderer';
import { View } from 'react-native';
import ExceptionCapture from '/common/ExceptionCapture';

describe('ExceptionCapture', () => {
  const Component = class TestComponent1 extends React.Component {
    render() {
      return <View style={{ backgroundColor: 'red' }} />;
    }
  };
  const ErrorComponent = class TestComponent2 extends React.Component {
    render() {
      return <View style={{ backgroundColor: 'green' }} {...this.props} />;
    }
  };
  const JSX = (
    <ExceptionCapture component={ErrorComponent}>
      <ExceptionCapture component={ErrorComponent}>
        <Component />
      </ExceptionCapture>
    </ExceptionCapture>
  );

  it('renders currectly', () => {
    const wrapper = renderer.create(JSX);
    expect(wrapper.toJSON()).toMatchSnapshot();

    const globalHandler = ErrorUtils.getGlobalHandler();
    globalHandler(new Error('abort'), true);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('set the original errorHandler when component is unmounted', () => {
    const originalErrorHandler = () => {};
    ErrorUtils.setGlobalHandler(originalErrorHandler);

    const wrapper = renderer.create(JSX);
    wrapper.toJSON();
    wrapper.unmount();

    expect(ErrorUtils.getGlobalHandler()).toEqual(originalErrorHandler);
  });
});
