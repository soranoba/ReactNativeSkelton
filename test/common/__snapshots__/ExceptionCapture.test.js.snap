// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExceptionCapture renders currectly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "red",
    }
  }
/>
`;

exports[`ExceptionCapture renders currectly 2`] = `
<View
  error={[Error: abort]}
  style={
    Object {
      "backgroundColor": "green",
    }
  }
/>
`;
